rust/gui-client/src-tauri/src/client/network_changes/windows.rs

```diff
@@ -422,3 +398,216 @@ fn get_apartment_type() -> WinResult<(Com::APTTYPE, Com::APTTYPEQUALIFIER)> {
+
+        rt.block_on(async move {
+            loop {
+                tokio::select! {
+                    _r = tokio::signal::ctrl_c() => break,
+                    r = listener_4.notified() => r?,
+                    r = listener_6.notified() => r?,
+                }
+
+                let resolvers = crate::client::resolvers::get()?;
+                tracing::info!(?resolvers);
+            }
+
+            listener_4.close()?;
+            listener_6.close()?;
+            Ok::<_, anyhow::Error>(())
+        })?;
+
+        Ok(())
+    }
+
+    pub(crate) struct CombinedListener {
+        listener_4: Listener,
+        listener_6: Listener,
+    }
+
+    impl CombinedListener {
+        pub(crate) fn new() -> Result<Self> {
+            let [key_ipv4, key_ipv6] = open_network_registry_keys()?;
+            let listener_4 = Listener::new(key_ipv4)?;
+            let listener_6 = Listener::new(key_ipv6)?;
+
+            Ok(Self {
+                listener_4,
+                listener_6,
+            })
+        }
+
+        pub(crate) fn close(&mut self) -> Result<()> {
+            self.listener_4.close()?;
+            self.listener_6.close()?;
+            tracing::debug!("Unregistered registry listener");
+            Ok(())
+        }
+
+        pub(crate) async fn notified(&mut self) -> Result<()> {
+            tokio::select! {
+                r = self.listener_4.notified() => r?,
+                r = self.listener_6.notified() => r?,
+            }
+            Ok(())
+        }
+    }
+
+    struct Listener {
+        key: winreg::RegKey,
+        notify: std::pin::Pin<Box<Notify>>,
+        wait_handle: Option<HANDLE>,
+        event: Option<HANDLE>,
+    }
+
+    impl Listener {
+        pub(crate) fn new(key: winreg::RegKey) -> Result<Self> {
+            let notify = Box::pin(Notify::default());
+            let notify_ptr: *const Notify = notify.deref();
+            let event = unsafe { CreateEventA(None, false, false, None) }?;
+            let mut wait_handle = HANDLE(0isize);
+
+            // The docs say that `RegisterWaitForSingleObject` uses "a worker thread" from
+            // "the thread pool".
+            // Raymond Chen, who is an authority on Windows internals, says RegisterWaitForSingleObject
+            // does multiple waits on the same worker thread internally:
+            // <https://devblogs.microsoft.com/oldnewthing/20081117-00/?p=20183>
+            // There is a different function in the CLR called `RegisterWaitForSingleObject`,
+            // so he might be talking about that.
+            // If he's not, then we're invisibly tying up two worker threads. Can't help it.
+
+            // SAFETY: It's complicated.
+            // The callback here can cause a lot of problems. We pin the `Notify` object.
+            // We don't use an `Arc` because `Notify` already has only `&self` methods, and
+            // the callback has no way to free an `Arc`.
```

Comment:
The callback can decrement the `Arc` using `Arc::from_raw`

Reply:
But when `drop` is called, we'll cancel the callback, which will forget the callback's cloned `Arc`, if the callback has not fired. We could check the `Receiver` to see whether we need to force the Arc to drop, but I thought it's simpler to imagine that the callback is taking a const borrow from `Listener`, and `Drop` enforces the lifetimes, and not that they're actually sharing ownership.

Reply:
but if we use the `arc` we don't need to unregister the callback, right?

Reply:
I don't understand. This one is the long-lived callback registered by `RegisterWaitForSingleObject` and unregistered in `Drop` by `UnregisterWaitEx`. If I didn't unregister the callback, it would leak the `Arc` and 1 or 2 Win32 handles.